Sources/Basics/VirtualFileSystem.swift

```diff
@@ -0,0 +1,225 @@
+        }
+    }
+
+    var isRoot: Bool {
+        switch self {
+        case .directory(_, _, _): return false
+        case .file(_, _, _): return false
+        case .root(_): return true
+        }
+    }
+
+    var isSymlink: Bool {
+        switch self {
+        case .directory(_, let isSymlink, _): return isSymlink
+        case .file(_, _, let isSymlink): return isSymlink
+        case .root(_): return false
+        }
+    }
+}
+
+private enum Errors: Swift.Error, LocalizedError {
+    case noSuchFileOrDirectory(path: AbsolutePath)
+    case notAFile(path: AbsolutePath)
+    case readOnlyFileSystem
+    case unhandledDirectoryNode(path: AbsolutePath)
+
+    public var errorDescription: String? {
+        switch self {
+        case .noSuchFileOrDirectory(let path): return "no such file or directory: \(path.pathString)"
+        case .notAFile(let path): return "not a file: \(path.pathString)"
+        case .readOnlyFileSystem: return "read-only filesystem"
+        case .unhandledDirectoryNode(let path): return "unhandled directory node: \(path.pathString)"
+        }
+    }
+}
+
+private extension FileSystem {
+    func getDirectoryNodes(_ path: AbsolutePath) throws -> [DirectoryNode] {
+        return try getDirectoryContents(path).compactMap {
+            let current = path.appending(component: $0)
+            let isSymlink = isSymlink(current)
+
+            if isFile(current) {
+                return .file(name: $0, isExecutable: isExecutableFile(current), isSymlink: isSymlink)
+            } else if isDirectory(current) {
+                if $0.hasPrefix(".") { return nil } // we ignore hidden files
+                return .directory(name: $0, isSymlink: isSymlink, children: try getDirectoryNodes(current))
+            } else {
+                throw Errors.unhandledDirectoryNode(path: current)
+            }
+        }
+    }
+}
+
+/// A JSON-backed, read-only virtual file system.
+public class VirtualFileSystem: FileSystem {
+    private let root: DirectoryNode
+    private let readFileContents: ((AbsolutePath) throws -> ByteString)
+
+    public init(path: AbsolutePath, fs: FileSystem, readFileContents: @escaping ((AbsolutePath) throws -> ByteString) = { _ in "" }) throws {
+        self.root = try JSONDecoder.makeWithDefaults().decode(path: path, fileSystem: fs, as: DirectoryNode.self)
+        self.readFileContents = readFileContents
+        assert(self.root.isRoot, "VFS needs to have a root node")
+    }
+
+    /// Write information about the directory tree at `directoryPath` into a JSON file at `vfsPath`. This can later be used to construct a `VirtualFileSystem` object.
+    public static func serializeDirectoryTree(_ directoryPath: AbsolutePath, into vfsPath: AbsolutePath, fs: FileSystem) throws {
+        let data = try JSONEncoder.makeWithDefaults().encode(DirectoryNode.root(children: fs.getDirectoryNodes(directoryPath)))
+        try data.write(to: URL(fileURLWithPath: vfsPath.pathString))
+    }
```

Comment:
maybe make this into an instance method `public func save(fileSystem: FileSystem, path: AbsolutePath)`?

Reply:
I don't really want to blur the line between `InMemoryFileSystem` and this too much, so the idea was that the only way you can obtain an instance is by reading from an existing file.